rfd/0002-custom-terraform-registry.md

```diff
@@ -0,0 +1,185 @@
+---
+authors: Trent Clarke (trent@goteleport.com)
+state: draft
+---
+
+# RFD 2 - Custom Terraform Registry
+
+## What
+
+A Terraform _registry_ is an index of available terraform _providers_. The
+Terraform tool uses registries to locate, validate and install Terraform
+providers for the various services required by terraform scripts. Hashicorp
+(the Terraform vendor) publishes the protocol that Terraform uses to interact
+with registries and so third parties (like us) may provide _custom_ Terraform
+registries outside the main, Hashicorp-provided public registry.
+
+This RFD investigates the use of a custom Terraform registry to distribute the
+Teleport provider for Terraform.
+
+## Why
+
+We have two main goals that effect how we offer our Terraform provider to
+customers
+
+ 1. We want to meet out customers where they are, e.g. if they're using
+    Terraform, we want to supply them with a Terraform provider, and a method
+    for installing it that is natural to the Terraform ecosystem.
+
+ 2. We want to maintain close control of the distribution of our software.
+
+With these two goals in mind, the most practical solution for distributing
+the Teleport Terraform provider appears to be hosting a custom Terraform
+registry.
+
+### Why not use the public Hashicorp registry?
+
+This conflicts with goal #2 (i.e. maintaining close control of the
+distribution of our software).
+
+### Will using a custom registry be onerous on our users?
+
+No. Certainly less onerous than the current solution (involving several
+manual download and installation steps).
+
+The user would reference the custom registry directly in their terraform
+scripts, and the Terraform tool would take care of the rest.
+
+```terraform
+terraform {
+    required_providers {
+        teleport = {
+            source = "https://goteleport.com/api/teleport/"
+            version = "9.1.0"
+        }
+    }
+}
+```
+
+For more info, see [the Terraform docs](https://www.terraform.io/cloud-docs/registry/using)
+on using custom registries.
+
+## Details
+
+### Terraform Registry Protocol requirements
+
+For the full specification, see the [Provider Registry Specification](https://www.terraform.io/internals/provider-registry-protocol).
+
+The major points of the registry protocol affecting this document are
+
+1. A web service that exposes the registry index,
+2. The binary packages referenced by the registry must be packaged in zipfiles,
+   as opposed to the compressed tarballs our current build process produces, and
+3. The checksum files must be signed with a GPG key
+
+### Serving the registry and packages
+
+#### Registry Hosting
+
+The prototype registry (under development) is currently set up to be served as
```

Comment:
Any reason we can't just host the registry index file off of S3 as well and update it as a part of the Drone publishing pipeline?